canonical-web-and-design/ubuntu.com
# Refactor purchase modal

## Done
Rewrote the modal as a react app using [react query](https://react-query.tanstack.com/), [Formik](https://formik.org/) and [React Vanilla components](https://canonical-web-and-design.github.io/react-components/?path=/docs/col--default-story)

The modal works the same way as before except for a couple of changes

## Changes
If a user is logged in and already registered a payment method with us, they won't have to go through the form again and instead directly land on the "recap / step 2" view 
![image](https://user-images.githubusercontent.com/11927929/120795021-3b156e00-c539-11eb-9220-6efb0a6073f9.png)


If a user is buying from a VAT applicable country and we already know their VAT number, we prepopulate alongside the rest of the information
![image](https://user-images.githubusercontent.com/11927929/120795218-7ca61900-c539-11eb-8bcd-aa62218bb24a.png)


If a user's account is in a locked state (because a purchase failed) we tell them to go to /payment-method to fix it 
![image](https://user-images.githubusercontent.com/11927929/120795709-1d94d400-c53a-11eb-9add-4dfbf23cf043.png)
(thanks @albertkol )

Removed the 45s timer after a purchase which was causing issues and was made and replaced it with 5 spaced retries.

## QA

https://ubuntu-com-9847.demos.haus/advantage/subscribe?test_backend=true

Logged in / logged out 

Few things to check:
- successful purchase with normal card (4242424242424242);
- successful purchase triggering 3ds modal with the card (4000002760003184)
- failed purchase 3ds modal with the card (4000002760003184) 
- other test cards can be found here: https://stripe.com/docs/testing
- as a guest use an existing email

## Issue / Card

Fixes
https://github.com/canonical-web-and-design/commercial-squad/issues/11
https://github.com/canonical-web-and-design/ubuntu.com/issues/9638
https://github.com/canonical-web-and-design/ubuntu.com/issues/9370
https://github.com/canonical-web-and-design/commercial-squad/issues/2
